### ISSUE TITLE
Failed Upgrade Causes Stale Operator Failure Status

### ISSUE THREAD

**Issue by johnrk** [cockroachdb/cockroach-operator#464]
For the [4/29 Bug bash](https://docs.google.com/spreadsheets/d/1mW7pMPXq9t7Rph9473xj3JmFXwQ769YQIzF668pTIWQ/edit#gid=139881396), I ran the Negative Testing Scenario

**Repro Steps:**
1. Install Operator
2. Using the Operator, install a version of CockroachDB v19.2.x
3. Using the Operator, attempt to upgrade to CockroachDB v20.2.x (expected to fail)
4. Using the Operator, attempt to install a version of CockroachDB v20.1.x (expected to succeed)

**Expected Results:** Step 3 is expected to fail. Step 4 should succeed, and getting status of the cluster via `kubectl describe CrdbCluster cockroachdb` should communicate that the cluster and the Operator are in a healthy status on v20.1.x.

**Actual Results:**
After installing v20.1.x in step 4, the kubectl describe status communicates:
- Cluster is in a failed state
- a stale message of a failed upgrade from step 3

**kubectl describe CrdbCluster cockroachdb:**
```
Status:
  Cluster Status:  Failed
  Conditions:
    Last Transition Time:  2021-04-29T14:15:48Z
    Status:                True
    Type:                  Initialized
    Last Transition Time:  2021-04-29T14:32:02Z
    Status:                True
    Type:                  CrdbVersionChecked
    Last Transition Time:  2021-04-29T14:14:57Z
    Status:                True
    Type:                  CertificateGenerated
  Crdbcontainerimage:      cockroachdb/cockroach:v20.2.7
  Operator Actions:
    Last Transition Time:  2021-04-29T14:15:48Z
    Message:               
    Status:                Finished
    Type:                  Initialize
    Last Transition Time:  2021-04-29T14:33:46Z
    Message:               failed to update sts with partitioned update: cockroachdb: upgrading from 20.1.11 to 20.2.7 with preserve downgrade option set to 19.2.0 not allowed: can't roll forward due to preserve downgrade option
    Status:                Failed
    Type:                  PartialUpdate
  Version:                 v20.2.7
Events:                    <none>
```

**Additional context:**
- On Macbook
- operator version: `image: cockroachdb/cockroach-operator:v1.7.5`

**Comment by alinadonisa** [comment]
@johnrk  upgrading from **20.1.11 to 20.2.7** with preserve downgrade option set to 19.2.0 not allowed: can't roll forward due to preserve downgrade option. This is a validation done when we are trying to upgrade from 20.1.11 to 20.2.7.  S o you managed to update to 20.1.11 and when you tried to update to 20.2.7 it failed?

**Comment by johnrk** [comment]
@alinadonisa , Based on my repro steps, I don't think I preserved downgrade option. 

The issue that I am trying to highlight is our failure to clear out old error messages in the status of the CrdbCluster instance.

**Comment by keith-mcclellan** [comment]
@johnrk I actually think @alinadonisa is correct and this is working as designed. The Operator will ALWAYS set the preserve downgrade option for a major upgrade, and it wasn't cleared by you so the upgrade isn't finalized. So when you then tried to upgrade to 20.2.x you got a good error. I agree that this is confusing (particularly `Cluster Status: failed`) when the database is actually healthy but the last action the operator tried to take against this CR did fail so I can see how we got to this state.  

@taroface can we make sure that we have explicit upgrade instructions for the k8s operator that clarify that we always set the preserve downgrade option and the user is required to manually finalize the upgrade?

Additionally, @chrislovecnm and @alinadonisa we should backlog an issue to make this less confusing - maybe by renaming `cluster status` to something like `last cluster action`?  @udnay would be great to get your thoughts here too

**Comment by taroface** [comment]
> @taroface can we make sure that we have explicit upgrade instructions for the k8s operator that clarify that we always set the preserve downgrade option and the user is required to manually finalize the upgrade?

The [current upgrade docs](https://www.cockroachlabs.com/docs/v20.2/orchestrate-cockroachdb-with-kubernetes.html#upgrade-the-cluster) instruct you to set `cluster.preserve_downgrade_option` and to manually finalize. 

Do you mean that:

- The Operator automatically sets `cluster.preserve_downgrade_option` to the previous major version?
- This means that there is never auto-finalization when using the Operator?

**Comment by udnay** [comment]
If the `Cluster Status` does not represent the actual state of the current cluster then I can see how it is confusing. Ideally the status is `Ready`, `Restarting`, `Initializing` etc to let people know if the cluster is ok or doing a specific workflow like certificate renewl or something. IMO the workflows are probably better tracked by having the operator emit events for "upgrade" or "cert renewal" or "decommissioning" etc and let the status be a quick check if the cluster is doing an operation or in stable state. 
wdyt?

**Comment by alinadonisa** [comment]
@udnay  now the status of the operator it is actually failed if one of the actions fails. It always take in consideration the last action it runs. It is not focus on the health of the cluster. I agree we should redesign.  We should  definitely record the events in k8s.  I added  issue https://github.com/cockroachdb/cockroach-operator/issues/490 for this.
@johnrk you can close this one.

**Comment by keith-mcclellan** [comment]
> @keith-mcclellan @alinadonisa
> 
> > @taroface can we make sure that we have explicit upgrade instructions for the k8s operator that clarify that we always set the preserve downgrade option and the user is required to manually finalize the upgrade?
> 
> The [current upgrade docs](https://www.cockroachlabs.com/docs/v20.2/orchestrate-cockroachdb-with-kubernetes.html#upgrade-the-cluster) instruct you to set `cluster.preserve_downgrade_option` and to manually finalize.
> 
> Do you mean that:
> 
> * The Operator automatically sets `cluster.preserve_downgrade_option` to the previous major version?

Yes, the operator does this automatically.  The user just needs to finalize (we don't auto-finalize)

> * This means that there is never auto-finalization when using the Operator?

That is correct.  We should backlog an issue to allow customers to auto-finalize for dev/test clusters.
